File: compiler/src/queries/TypeQueries.swift
extension Queries {
    public static var hasType: Query<ConstructedType> {
        { db, node, body in
            guard let group = db[node, Typed.self]?.group, let type = group.types.first else {
                return
            }

            body(type)
        }
    }

    public static var inGroup: Query<Node> {
        { db, node, body in
            guard let group = db[node, Typed.self]?.group else { return }

            for other in group.nodes { body(other) }
        }
    }

    public struct ConflictingTypes {
        let source: Node?
        let from: Node
        let nodes: [Node]
        let types: [ConstructedType]
        let trace: [Constraint]
    }

    public static var conflictingTypes: Query<ConflictingTypes> {
        { db, node, body in
            guard let group = db[node, Typed.self]?.group, group.types.count > 1,
                group.nodes.first == node
            else { return }

            let source = db[node, Instantiated.self]?.sourceNode
            let nodes = group.nodes.filter { $0 != node }

            body(
                .init(
                    source: source,
                    from: node,
                    nodes: nodes,
                    types: group.types,
                    trace: group.trace,
                )
            )
        }
    }

    public static var incompleteType: Query<Node, ConstructedType> {
        { db, node, body in
            guard let group = db[node, Typed.self]?.group, group.types.count == 1,
                group.nodes.first == node
            else { return }

            let type = group.types[0]

            guard Type.constructed(type).referencesNodes else { return }

            body(node, type)
        }
    }

    public static var unknownType: Query<> {
        { db, node, body in
            guard let typed = db[node, Typed.self] else { return }

            guard let group = typed.group else {
                body()
                return
            }

            guard group.types.isEmpty, group.nodes.first == node else { return }

            body()
        }
    }
}

File: compiler/queries/src/type_queries.rs
use crate::QueryCtx;
use std::collections::BTreeSet;
use wipple_core::{
    db::Node,
    traces::Traces,
    typecheck::{
        groups::{NodeRank, Typed},
        instantiate::Instantiated,
        ty::ConstructedTy,
    },
};

pub fn has_type<'a>(db: &QueryCtx<'a>, node: Node) -> Option<&'a ConstructedTy> {
    let Typed(Some(group)) = db.get(node)? else {
        return None;
    };

    group.tys().next()
}

pub fn in_group(db: &QueryCtx<'_>, node: Node) -> impl Iterator<Item = Node> {
    let Some(Typed(Some(group))) = db.get(node) else {
        return Default::default();
    };

    group.nodes().collect::<Vec<_>>().into_iter()
}

#[derive(Debug, Clone)]
pub struct ConflictingTypes {
    pub source: Option<Node>,
    pub from: Node,
    pub related: BTreeSet<Node>,
    pub group: BTreeSet<Node>,
    pub tys: Vec<ConstructedTy>,
    pub traces: Traces,
}

pub fn conflicting_types(db: &QueryCtx<'_>, node: Node) -> Option<ConflictingTypes> {
    let Typed(Some(group)) = db.get(node)? else {
        return None;
    };

    if group.tys().count() <= 1 {
        return None;
    }

    if group.get_rank(node) > group.min_rank() {
        return None;
    }

    let traces = db.traces_for(node, group.nodes());

    let source = db
        .get::<Instantiated>(node)
        .map(|instantiated| instantiated.source_node);

    let mut related = group.nodes().collect::<BTreeSet<_>>();
    related.remove(&node);
    related.retain(|&node| group.get_rank(node) <= NodeRank::Inherited);

    Some(ConflictingTypes {
        source,
        from: node,
        related,
        group: group.nodes().collect(),
        tys: group.tys().cloned().collect(),
        traces,
    })
}

pub fn incomplete_type<'a>(db: &QueryCtx<'a>, node: Node) -> Option<(Node, &'a ConstructedTy)> {
    let Typed(Some(group)) = db.get(node)? else {
        return None;
    };

    let mut tys = group.tys();
    let ty = tys.next()?;

    if tys.next().is_some() {
        return None;
    }

    if ty.children.iter().any(|&ty| {
        db.get(ty)
            .and_then(|Typed(group)| group.as_ref())
            .is_some_and(|group| group.tys().next().is_none())
    }) {
        return Some((node, ty));
    }

    None
}

pub fn unknown_type(db: &QueryCtx<'_>, node: Node) -> bool {
    let Some(Typed(group)) = db.get(node) else {
        return false;
    };

    let Some(group) = group else {
        return true;
    };

    group.tys().next().is_none()
}
